core/controllers/reader.py

```diff
@@ -1274,3 +1299,103 @@ def put(self, entity_type, entity_id):
             entity_type, state_reference,
             interaction_id, answer, answer_details)
         self.render_json({})
+
+
+class CheckpointCompletedEventHandler(base.BaseHandler):
+    """Tracks a learner completing a checkpoint."""
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+            'schema': editor.SCHEMA_FOR_EXPLORATION_ID
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'PUT': {
+            'last_completed_checkpoint_exp_version': {
+                'schema': editor.SCHEMA_FOR_VERSION
+            },
+            'last_completed_checkpoint_state_name': {
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'has_length_at_most',
+                        'max_value': constants.MAX_STATE_NAME_LENGTH
+                    }]
+                }
+            },
+        }
+    }
+
+    REQUIRE_PAYLOAD_CSRF_CHECK = False
+
+    @acl_decorators.can_play_exploration
+    def put(self, exploration_id):
+        """Handles PUT requests.
+
+        Args:
+            exploration_id: str. The ID of the exploration.
+        """
+
+        user_id = self.user_id
+        last_completed_checkpoint_state_name = self.normalized_payload.get(
+            'last_completed_checkpoint_state_name')
+        last_completed_checkpoint_exp_version = self.normalized_payload.get(
+            'last_completed_checkpoint_exp_version')
+
+        user_services.update_learner_checkpoint_progress(
+            user_id,
+            exploration_id,
+            last_completed_checkpoint_state_name,
+            last_completed_checkpoint_exp_version)
+
+        self.render_json(self.values)
+
+
+class CheckpointVisitedEventHandler(base.BaseHandler):
+    """Tracks a learner visiting a previously completed checkpoint."""
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+            'schema': editor.SCHEMA_FOR_EXPLORATION_ID
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'PUT': {
+            'last_completed_checkpoint_exp_version': {
```

Comment:
Should this be latest_completed_checkpoint_exp_version or just exploration_version? It seems odd to refer to "last completed" in a "latest visited" handler.

Reply:
Actually, this field is meant to store the exploration version where the learner has last made any progress. Suppose the learner last made any progress (i.e., visited or completed a checkpoint in version x) and the editor updates the exploration to be of version x+1. Then `version` property of exploration dict would store x+1 and this property would still store x. Do you have any suggestions on how this variable should be named?

The context on why we are updating this field on `checkpoint_visited_event` can be found [here](https://github.com/oppia/oppia/pull/15213#discussion_r835750238).

Reply:
The exploration version needs to be there so that it corresponds to the state name for the last/latest checkpoint. The difficulty here is not storing the exploration version; the issue here is the confusion between "last" and "latest". Does that clarify my comment?

Also, I think we unfortunately still need to figure out more descriptive names for "last" and "latest". I still can't remember which is which, when reviewing, so other people will probably get confused as well.

Reply:
Hmm, you are right about the literal meanings and I think it would make sense to use them instead of 'last' and 'latest', so for `last_completed_checkpoint_state_name`, we can use `furthest_completed_checkpoint_state_name` and for `latest_viewed_checkpoint_state_name` we could use `most_recently_viewed_checkpoint_state_name`, as for exploration version, we could use saved_checkpoints_progress_exp_version for better clarity and naming it seperate from the concept of most recent and furthest.

Reply:
That definitely sounds better! Thanks @paprajapati9

Reply:
Thanks @paprajapati9, the names for the state names sound good.

I am confused about the version point, though. Suppose a learner goes all the way to the fifth checkpoint (when the lesson is at v70), and then, after a while when the lesson is already at v71, restarts and plays the first checkpoint. Wouldn't the versions for "furthest completed" and "most recently viewed" be different in that case?

Reply:
@seanlip good point, this was something to consider for versioning, I was thinking to deal with versions, every time inside the get request handler, we check the version of the saved progress's exploration with the version of the current exploration and if there is a mismatch, there can be 5 scenerios:

1. The most recent and furthest checkpoints are still present in this version, in this case we just update the progress's version of exploration and move on.
2. Most recent checkpoint was removed but furthest is still present, in this case we set the most recent to equal the furthest checkpoint and update saved progress exp version
3. Furthest is deleted but most recent is present, in this case we check the checkpoint in the saved progress before the furthest that is present in the current version and set it as the furthest and update exploration version
4. If both furthest and moat recent are deleted, in that case, we like the last case check for the checkpoint in the saved progress exp version present in the current exp version and set it as furthest and the most recent and update saved progress exp version.
5. An edge case of case 4 os when no such checkpoint is found, in that case, we delete the corresponding UserExplorationModel and treat it like no progress was stored.

Basically, I think we should not overly complicate things by storing seperate exp versions for furthest and moat recent checkpoints and keep it single viewing it as saved progress's exp version. Thoughts ?

Reply:
But the problem is that a state name is tied to a version and doesn't make sense without it, right? E.g. versions 70 and 71 may have different sets of state names due to state additions and deletions, so really you're dealing with a (version, state name) tuple conceptually.

Reply:
The variables have now been renamed and the working has been updated as per our discussion. PTAL! Thanks